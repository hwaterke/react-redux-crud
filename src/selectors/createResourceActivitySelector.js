// @flow
import type {ActivityState} from '../reducers/createActivityReducerForResource'
import type {ResourceDefinition} from '../types/ResourceDefinition'
import type {Selector} from '../types/Selector'

export type ActivitySelector = any => ActivityState
export type ActivityResourceSelector = ResourceDefinition => ActivitySelector

export function createResourceActivitySelector(
  stateLocationSelector: Selector
): ActivityResourceSelector {
  const loadingSelectorRegistry = {}

  return function(resource: ResourceDefinition): ActivitySelector {
    if (!loadingSelectorRegistry[resource.name]) {
      loadingSelectorRegistry[resource.name] = state =>
        stateLocationSelector(state)[resource.name]
    }

    return loadingSelectorRegistry[resource.name]
  }
}
