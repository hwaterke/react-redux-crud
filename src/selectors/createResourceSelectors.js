// @flow
import {createSelector} from 'reselect'
import type {ResourceDefinition} from '../types/ResourceDefinition'
import type {Selector} from '../types/Selector'

export type ResourceSelector = any => {[string]: any}
export type ResourceResourceSelector = ResourceDefinition => ResourceSelector
export type ResourceSelectors = {
  byId: ResourceResourceSelector,
  asArray: ResourceResourceSelector,
}

export function createResourceSelectors(
  stateLocationSelector: Selector
): ResourceSelectors {
  const byIdSelectorRegistry = {}
  const arraySelectorRegistry = {}

  const byId = function(resource: ResourceDefinition) {
    if (!byIdSelectorRegistry[resource.name]) {
      byIdSelectorRegistry[resource.name] = state =>
        stateLocationSelector(state)[resource.name]
    }

    return byIdSelectorRegistry[resource.name]
  }

  const asArray = function(resource: ResourceDefinition) {
    if (!arraySelectorRegistry[resource.name]) {
      arraySelectorRegistry[resource.name] = createSelector(
        byId(resource),
        resourceById => Object.values(resourceById)
      )
    }

    return arraySelectorRegistry[resource.name]
  }

  return {
    byId,
    asArray,
  }
}
