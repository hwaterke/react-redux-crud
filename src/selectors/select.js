// @flow
import {createSelector} from 'reselect'
import type {ResourceDefinition} from '../types/ResourceDefinition'

type Selectors = {
  byId: Function,
  asArray: Function,
  activity: {
    all: Function,
    totalFetching: Function,
    totalCreating: Function,
    totalUpdating: Function,
    totalRemoving: Function,
  },
}

const selectorsByResource = {}

export function select(resource: ResourceDefinition): Selectors {
  const {name, reduxResourceSelector, reduxResourceActivitySelector} = resource

  if (!selectorsByResource[name]) {
    // Create a new set of selectors
    const selectors = {}

    if (reduxResourceSelector) {
      // Specific resource location specified ?
      selectors.byId = state => reduxResourceSelector(state)[name]
    } else {
      // Default location
      selectors.byId = state => state.resources[name]
    }

    selectors.activity = {}
    if (reduxResourceActivitySelector) {
      // Specific resource location specified ?
      selectors.activity.all = state =>
        reduxResourceActivitySelector(state)[name]
    } else {
      // Default location
      selectors.activity.all = state => state.resourcesActivity[name]
    }

    selectors.asArray = createSelector(selectors.byId, resourceById =>
      Object.values(resourceById)
    )

    selectors.activity.totalFetching = createSelector(
      selectors.activity.all,
      activity => Object.keys(activity.fetching).length
    )

    selectors.activity.totalCreating = createSelector(
      selectors.activity.all,
      activity => Object.keys(activity.creating).length
    )

    selectors.activity.totalUpdating = createSelector(
      selectors.activity.all,
      activity => Object.keys(activity.updating).length
    )

    selectors.activity.totalRemoving = createSelector(
      selectors.activity.all,
      activity => Object.keys(activity.removing).length
    )

    // Store selector instances in global object
    selectorsByResource[name] = selectors
  }

  // Return existing selectors
  return selectorsByResource[name]
}
