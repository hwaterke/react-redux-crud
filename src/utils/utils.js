// @flow

export function invariant(condition: any, message: string) {
  if (!condition) {
    const error = new Error(message)
    error.name = 'Invariant Violation'
    throw error
  }
}
