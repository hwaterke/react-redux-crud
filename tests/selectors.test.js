// @flow
import reduxCrud from 'redux-crud/dist/index'
import type {ActivityResourceSelector} from '../src/selectors/createResourceActivitySelector'
import {createResourceActivitySelector} from '../src/selectors/createResourceActivitySelector'
import {createResourceSelectors} from '../src/selectors/createResourceSelectors'
import {Book, initialiseStore} from './utils'

describe('Selectors', () => {
  let store

  beforeEach(() => {
    store = initialiseStore()
  })

  describe('Resources', () => {
    let selector

    beforeEach(() => {
      selector = createResourceSelectors(state => state.resources)
    })

    describe('byId', () => {
      test('should have a correct initial state', function() {
        const state = selector.byId(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })

      test('should have a correct state after FETCH_START', function() {
        const actionTypes = reduxCrud.actionTypesFor(Book.name)
        store.dispatch({type: actionTypes.fetchStart})

        const state = selector.byId(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })

      test('should have a correct state after FETCH_SUCCESS', function() {
        const actionTypes = reduxCrud.actionTypesFor(Book.name)
        store.dispatch({type: actionTypes.fetchStart})
        store.dispatch({
          type: actionTypes.fetchSuccess,
          records: [{uuid: 1, name: 'Name'}],
        })

        const state = selector.byId(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })
    })

    describe('asArray', () => {
      test('should have a correct initial state', function() {
        const state = selector.asArray(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })

      test('should have a correct state after FETCH_START', function() {
        const actionTypes = reduxCrud.actionTypesFor(Book.name)
        store.dispatch({type: actionTypes.fetchStart})

        const state = selector.asArray(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })

      test('should have a correct state after FETCH_SUCCESS', function() {
        const actionTypes = reduxCrud.actionTypesFor(Book.name)
        store.dispatch({type: actionTypes.fetchStart})
        store.dispatch({
          type: actionTypes.fetchSuccess,
          records: [{uuid: 1, name: 'Name'}],
        })

        const state = selector.asArray(Book)(store.getState())
        expect(state).toMatchSnapshot()
      })
    })
  })

  describe('Activity', () => {
    let selector: ActivityResourceSelector

    beforeEach(() => {
      selector = createResourceActivitySelector(state => state.activity)
    })

    test('should have a correct initial state', function() {
      const state = selector(Book)(store.getState())
      expect(state).toMatchSnapshot()
    })

    test('should have a correct state after FETCH_START', function() {
      const actionTypes = reduxCrud.actionTypesFor(Book.name)
      store.dispatch({type: actionTypes.fetchStart})

      const state = selector(Book)(store.getState())
      expect(state).toMatchSnapshot()
    })

    test('should have a correct state after FETCH_SUCCESS', function() {
      const actionTypes = reduxCrud.actionTypesFor(Book.name)
      store.dispatch({type: actionTypes.fetchStart})
      store.dispatch({
        type: actionTypes.fetchSuccess,
        records: [{uuid: 1, name: 'Name'}],
      })

      const state = selector(Book)(store.getState())
      expect(state).toMatchSnapshot()
    })
  })
})
