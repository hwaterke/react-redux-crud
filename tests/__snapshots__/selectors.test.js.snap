// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors Activity should have a correct initial state 1`] = `
Object {
  "activity": 0,
  "create": 0,
  "fetch": 0,
  "remove": 0,
  "update": 0,
}
`;

exports[`Selectors Activity should have a correct state after FETCH_START 1`] = `
Object {
  "activity": 1,
  "create": 0,
  "fetch": 1,
  "remove": 0,
  "update": 0,
}
`;

exports[`Selectors Activity should have a correct state after FETCH_SUCCESS 1`] = `
Object {
  "activity": 0,
  "create": 0,
  "fetch": 0,
  "remove": 0,
  "update": 0,
}
`;

exports[`Selectors Resources asArray should have a correct initial state 1`] = `Array []`;

exports[`Selectors Resources asArray should have a correct state after FETCH_START 1`] = `Array []`;

exports[`Selectors Resources asArray should have a correct state after FETCH_SUCCESS 1`] = `
Array [
  Object {
    "name": "Name",
    "uuid": 1,
  },
]
`;

exports[`Selectors Resources byId should have a correct initial state 1`] = `Object {}`;

exports[`Selectors Resources byId should have a correct state after FETCH_START 1`] = `Object {}`;

exports[`Selectors Resources byId should have a correct state after FETCH_SUCCESS 1`] = `
Object {
  "1": Object {
    "name": "Name",
    "uuid": 1,
  },
}
`;
